source/predicates/predicate.ts

```diff
@@ -86,13 +86,15 @@ export class Predicate<T = unknown> implements BasePredicate<T> {
 	/**
 	@hidden
 	*/
-	[testSymbol](value: T, main: Main, label: string | Function): asserts value {
+	[testSymbol](value: T | undefined, main: Main, label: string | Function): asserts value {
 		for (const {validator, message} of this.context.validators) {
 			if (this.options.optional === true && value === undefined) {
 				continue;
 			}
 
-			const result = validator(value);
+			const knownValue = value as T;
```

Comment:
Do you really have to use `as` casting here? TS should in theory be able to infer it correctly as we check for undefined on line 91.

Reply:
That casting is required since without a conditional, I don't think TS will infer the type. I think if we enclose the following code in `value !== undefined` and use `else { continue }`, type inference will work the way you expect.

I'll give you a few options soon when I test a few conditionals

Reply:
> I'll give you a few options soon when I test a few conditionals

Bump :)

Reply:
Big oops, forgot this was pending

I think this De Morganed version of the conditional explains the problem best - 
![image](https://user-images.githubusercontent.com/8168401/94654460-97484280-0359-11eb-8cdb-cbb326096ec7.png)

The first condition might be true and allow access to the inner scope, so TS is not able to infer that the value is indeed not undefined.

So we need to handle the case where it is optional not and undefined explicitly... how do you want it to work?

If the ArgumentError is to be thrown, that needs to be modified to take `T | undefined` type

Reply:
Right. It's better to use `value!` over `value as T` though, as the former is more type-safe.

Reply:
Do you just want me to change the cast or do you want me to change the input type of that validator function?

Reply:
Just the cast.